### PR TITLE
opencoarrays: add a test case to check the open-mpi version

### LINKED ISSUE #110650: opencoarrays installation out of sync with open-mpi dependency version
### `brew config` output

```shell
HOMEBREW_VERSION: 3.6.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6e2b162c4786e075323f038d46bfb566d91889e7
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 4f38944f7f5afeda86609f8bcb7b944400fc6025
Core tap last commit: 4 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 13.1.6 build 1316
Git: 2.32.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.5.1-x86_64
CLT: 13.4.0.0.1.1651278267
Xcode: 14.0
```

### `brew doctor` output

```shell
Your system is ready to brew.
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I'm trying to install OpenCoarrays via `brew install opencoarrays`, which installs the prerequisite package open-mpi 4.1.4_2.

### What happened (include all command output)?

The OpenCoarrays installation includes a `caf` script that invokes `gfortran` with various flags,  including one flag that attempts to link open-mpi as follows: `-L/usr/local/Cellar/open-mpi/4.1.4_1/lib`, which doesn't exist and doesn't match `brew`-installed open-mpi version. The following command output shows what the `caf` script is trying to do:
```
% caf --show
/usr/local/bin/gfortran -I/usr/local/Cellar/opencoarrays/2.10.0_1/include/OpenCoarrays-2.10.0_GNU-12.1.0 -fcoarray=lib -Wl,-flat_namespace -Wl,-commons,use_dylibs -L/usr/local/Cellar/open-mpi/4.1.4_1/lib -L/usr/local/Cellar/libevent/2.1.12/lib ${@} /usr/local/Cellar/opencoarrays/2.10.0_1/lib/libcaf_mpi.a /usr/local/lib/libmpi_usempif08.dylib /usr/local/lib/libmpi_usempi_ignore_tkr.dylib /usr/local/lib/libmpi_mpifh.dylib /usr/local/lib/libmpi.dylib
```

### What did you expect to happen?

I expected the `brew`-installed opencoarrays to use the version the `brew`-installed open-mpi.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew update
brew install opencoarrays
```
```

### PR DESCRIPTION
**Pull request by miles170**
Closes #110650.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/opencoarrays.rb [hunk 1]
```diff
@@ -51,5 +51,7 @@ def install
     EOS
     system "#{bin}/caf", "tally.f90", "-o", "tally"
     system "#{bin}/cafrun", "-np", "3", "--oversubscribe", "./tally"
+    %r{/open-mpi/(?<openmpi_version>[\d._]*)/lib} =~ shell_output("#{bin}/caf --show")
+    assert_equal openmpi_version, Formula["open-mpi"].pkg_version.to_s
```
Comment (SMillerDev) [inline Formula/opencoarrays.rb:55]: How about this?
```suggestion
    assert_match Formula["open-mpi"].version.to_s, shell_output("#{bin}/caf --show")
```

Comment (miles170) [reply Formula/opencoarrays.rb:55]: I think it should be `pkg_version` to compare `version` and `revision` at the same time.

Comment (SMillerDev) [reply Formula/opencoarrays.rb:55]: Maybe this is better then
```suggestion
    assert_match Formula["open-mpi"].prefix, shell_output("#{bin}/caf --show")
```

Comment (miles170) [reply Formula/opencoarrays.rb:55]: `Formula["open-mpi"].prefix` would return `/home/linuxbrew/.linuxbrew/opt/open-mpi`

`caf --show` uses `/home/linuxbrew/.linuxbrew/Cellar/open-mpi/4.1.4_1/lib`

Comment (miles170) [reply Formula/opencoarrays.rb:55]: Unless we can call `Formula["open-mpi"].versioned_prefix`, which is a private method.

Comment (SMillerDev) [reply Formula/opencoarrays.rb:55]: it doesn't for me, that would be `Formula["open-mpi"].opt_prefix`
```console
irb(main):001:0> Formula["open-mpi"].prefix
=> #<Pathname:/usr/local/Cellar/open-mpi/4.1.4_2>
```

We can also use `Formula["open-mpi"].lib`

Comment (miles170) [reply Formula/opencoarrays.rb:55]: ## ghcr.io/homebrew/ubuntu22.04:master

```
irb(main):001:0> Formula["open-mpi"].prefix
=> #<Pathname:/home/linuxbrew/.linuxbrew/opt/open-mpi>
irb(main):002:0> Formula["open-mpi"].lib
=> #<Pathname:/home/linuxbrew/.linuxbrew/opt/open-mpi/lib>
```
## brew config
```
HOMEBREW_VERSION: 3.6.1-33-g65da808
ORIGIN: https://github.com/Homebrew/brew
HEAD: 65da808a104d1cab0a6d1c5dde46b6ce85d60696
Last commit: 22 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: d0bc876efc548fd780a06b218b77ada39ad1c25c
Core tap last commit: 73 minutes ago
Core tap branch: opencoarrays
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DEVELOPER: set
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_AUTO_UPDATE: set
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: octa-core 64-bit unknown_0x6_0x8c
Clang: N/A
Git: 2.37.3 => /bin/git
Curl: 7.81.0 => /bin/curl
Kernel: Linux 5.15.0-47-generic x86_64 GNU/Linux
OS: Unknown
Host glibc: 2.35
/usr/bin/gcc: 11.2.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
xorg: N/A
```

Comment (SMillerDev) [reply Formula/opencoarrays.rb:55]: @Homebrew/brew is this intentional?

Comment (carlocab) [reply Formula/opencoarrays.rb:55]: > @Homebrew/brew is this intentional?

Yes.

Do
```ruby
Formula["open-mpi"].prefix.realpath
```

**Review by carlocab — APPROVED** [review]
Nice work @miles170!

**Pull request closed without merging**